File: remover.java
// remover.java
import java.io.*;
import java.nio.file.*;
import java.nio.file.attribute.*;
import java.util.*;
import java.util.regex.Pattern;

public class remover {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static boolean isIgnored(String name, Set<String> ignoreSet) {
        return ignoreSet.contains(name);
    }

    private static boolean isFolderEmpty(Path dir, Set<String> ignoreSet, long minSize) throws IOException {
        if (!Files.isDirectory(dir)) return false;
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir)) {
            List<Path> entries = new ArrayList<>();
            for (Path entry : stream) entries.add(entry);
            List<Path> filtered = new ArrayList<>();
            for (Path e : entries) {
                if (!isIgnored(e.getFileName().toString(), ignoreSet))
                    filtered.add(e);
            }
            if (filtered.isEmpty()) return true;
            if (minSize > 0) {
                boolean allSmall = true;
                for (Path e : filtered) {
                    if (Files.isDirectory(e)) { allSmall = false; break; }
                    if (Files.size(e) >= minSize) { allSmall = false; break; }
                }
                if (allSmall) return true;
            }
            return false;
        }
    }

    private static List<Path> getEmptyDirs(Path root, boolean recursive, List<String> exclude,
                                           Set<String> ignoreSet, long minSize) throws IOException {
        List<Path> result = new ArrayList<>();
        Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
            @Override
            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
                if (!recursive && !dir.equals(root)) return FileVisitResult.SKIP_SUBTREE;
                // Исключения
                for (String pat : exclude) {
                    if (Pattern.matches(pat, dir.getFileName().toString())) {
                        return FileVisitResult.SKIP_SUBTREE;
                    }
                }
                try {
                    if (isFolderEmpty(dir, ignoreSet, minSize)) {
                        result.add(dir);
                    }
                } catch (IOException ignored) {}
                return FileVisitResult.CONTINUE;
            }
        });
        // Сортировка по глубине (сначала глубокие)
        result.sort((a, b) -> b.getNameCount() - a.getNameCount());
        return result;
    }

    public static void main(String[] args) throws IOException {
        String root = ".";
        boolean recursive = true, dryRun = false, verbose = false, yes = false;
        String logFile = null;
        List<String> exclude = new ArrayList<>();
        List<String> ignoreList = new ArrayList<>();
        long minSize = 0;

        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            if (arg.equals("-p") && i+1 < args.length) root = args[++i];
            else if (arg.equals("-r")) recursive = true;
            else if (arg.equals("-n")) dryRun = true;
            else if (arg.equals("-v")) verbose = true;
            else if (arg.equals("-l") && i+1 < args.length) logFile = args[++i];
            else if (arg.equals("-e") && i+1 < args.length) exclude.addAll(Arrays.asList(args[++i].split(",")));
            else if (arg.equals("-i") && i+1 < args.length) ignoreList.addAll(Arrays.asList(args[++i].split(",")));
            else if (arg.equals("-m") && i+1 < args.length) minSize = Long.parseLong(args[++i]);
            else if (arg.equals("-y")) yes = true;
            else if (arg.equals("-h")) { System.out.println("Help..."); return; }
        }

        Path rootPath = Paths.get(root).toAbsolutePath();
        if (!Files.isDirectory(rootPath)) {
            System.err.println(colorize("Ошибка: '" + root + "' не является папкой", RED));
            System.exit(1);
        }

        Set<String> defaultIgnore = new HashSet<>(Arrays.asList(".DS_Store", "Thumbs.db", "desktop.ini"));
        defaultIgnore.addAll(ignoreList);
        Set<String> ignoreSet = defaultIgnore;

        List<Path> emptyDirs = getEmptyDirs(rootPath, recursive, exclude, ignoreSet, minSize);
        if (emptyDirs.isEmpty()) {
            System.out.println(colorize("Пустых папок не найдено.", YELLOW));
            return;
        }

        System.out.println(colorize("Найдено " + emptyDirs.size() + " пустых папок:", BLUE));
        if (verbose) {
            for (Path d : emptyDirs) System.out.println("  " + d);
        }

        if (dryRun) {
            System.out.println(colorize("Режим симуляции. Ничего не удалено.", GREEN));
            return;
        }

        if (!yes) {
            System.out.print(colorize("Удалить " + emptyDirs.size() + " папок? [y/N] ", YELLOW));
            try (Scanner sc = new Scanner(System.in)) {
                String ans = sc.nextLine();
                if (!ans.equalsIgnoreCase("y")) {
                    System.out.println(colorize("Операция отменена.", RED));
                    return;
                }
            }
        }

        List<String> logLines = new ArrayList<>();
        int deleted = 0;
        for (Path d : emptyDirs) {
            if (verbose) System.out.println(colorize("Удаление: " + d, GREEN));
            try {
                Files.walkFileTree(d, new SimpleFileVisitor<Path>() {
                    @Override
                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                        Files.delete(file);
                        return FileVisitResult.CONTINUE;
                    }
                    @Override
                    public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
                        Files.delete(dir);
                        return FileVisitResult.CONTINUE;
                    }
                });
                deleted++;
                logLines.add(d.toString());
            } catch (IOException e) {
                System.err.println(colorize("Ошибка удаления " + d + ": " + e.getMessage(), RED));
            }
        }

        if (logFile != null) {
            try {
                Files.write(Paths.get(logFile), logLines);
                System.out.println(colorize("Лог сохранён в " + logFile, GREEN));
            } catch (IOException e) {
                System.err.println(colorize("Ошибка записи лога: " + e.getMessage(), RED));
            }
        }

        System.out.println(colorize("Удалено " + deleted + " папок.", GREEN));
    }
}

File: remover.cs
// remover.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

class Remover
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static bool IsIgnored(string name, HashSet<string> ignoreSet)
    {
        return ignoreSet.Contains(name);
    }

    static bool IsFolderEmpty(string dir, HashSet<string> ignoreSet, long minSize)
    {
        if (!Directory.Exists(dir)) return false;
        var entries = Directory.GetFileSystemEntries(dir);
        var filtered = entries.Where(f => !ignoreSet.Contains(Path.GetFileName(f))).ToList();
        if (filtered.Count == 0) return true;
        if (minSize > 0)
        {
            bool allSmall = true;
            foreach (var f in filtered)
            {
                if (Directory.Exists(f))
                {
                    allSmall = false;
                    break;
                }
                var fi = new FileInfo(f);
                if (fi.Length >= minSize)
                {
                    allSmall = false;
                    break;
                }
            }
            if (allSmall) return true;
        }
        return false;
    }

    static List<string> GetEmptyDirs(string root, bool recursive, List<string> exclude, HashSet<string> ignoreSet, long minSize)
    {
        var result = new List<string>();
        var dirs = new Stack<string>();
        dirs.Push(root);
        while (dirs.Count > 0)
        {
            string current = dirs.Pop();
            try
            {
                var sub = Directory.GetDirectories(current);
                foreach (var d in sub)
                {
                    if (recursive) dirs.Push(d);
                    // Исключения
                    bool skip = false;
                    foreach (var pat in exclude)
                    {
                        if (System.Text.RegularExpressions.Regex.IsMatch(Path.GetFileName(d), WildcardToRegex(pat)))
                        {
                            skip = true;
                            break;
                        }
                    }
                    if (skip) continue;
                    if (IsFolderEmpty(d, ignoreSet, minSize))
                    {
                        result.Add(d);
                    }
                }
            }
            catch (Exception) { }
        }
        // Сортировка по глубине
        result.Sort((a, b) => b.Count(c => c == Path.DirectorySeparatorChar) - a.Count(c => c == Path.DirectorySeparatorChar));
        return result;
    }

    static string WildcardToRegex(string pattern)
    {
        return "^" + Regex.Escape(pattern).Replace("\\*", ".*").Replace("\\?", ".") + "$";
    }

    static void Main(string[] args)
    {
        string root = ".";
        bool recursive = true, dryRun = false, verbose = false, yes = false;
        string logFile = null;
        var exclude = new List<string>();
        var ignoreList = new List<string>();
        long minSize = 0;

        for (int i = 0; i < args.Length; i++)
        {
            string arg = args[i];
            if (arg == "-p" && i+1 < args.Length) root = args[++i];
            else if (arg == "-r") recursive = true;
            else if (arg == "-n") dryRun = true;
            else if (arg == "-v") verbose = true;
            else if (arg == "-l" && i+1 < args.Length) logFile = args[++i];
            else if (arg == "-e" && i+1 < args.Length) exclude.AddRange(args[++i].Split(','));
            else if (arg == "-i" && i+1 < args.Length) ignoreList.AddRange(args[++i].Split(','));
            else if (arg == "-m" && i+1 < args.Length) minSize = long.Parse(args[++i]);
            else if (arg == "-y") yes = true;
            else if (arg == "-h") { Console.WriteLine("Help..."); return; }
        }

        if (!Directory.Exists(root))
        {
            Console.WriteLine(Colorize($"Ошибка: '{root}' не является папкой", "red"));
            return;
        }

        var defaultIgnore = new HashSet<string> { ".DS_Store", "Thumbs.db", "desktop.ini" };
        foreach (var s in ignoreList) defaultIgnore.Add(s);
        var ignoreSet = defaultIgnore;

        var emptyDirs = GetEmptyDirs(root, recursive, exclude, ignoreSet, minSize);
        if (emptyDirs.Count == 0)
        {
            Console.WriteLine(Colorize("Пустых папок не найдено.", "yellow"));
            return;
        }

        Console.WriteLine(Colorize($"Найдено {emptyDirs.Count} пустых папок:", "blue"));
        if (verbose) emptyDirs.ForEach(d => Console.WriteLine($"  {d}"));

        if (dryRun)
        {
            Console.WriteLine(Colorize("Режим симуляции. Ничего не удалено.", "green"));
            return;
        }

        if (!yes)
        {
            Console.Write(Colorize($"Удалить {emptyDirs.Count} папок? [y/N] ", "yellow"));
            string ans = Console.ReadLine();
            if (ans?.ToLower() != "y")
            {
                Console.WriteLine(Colorize("Операция отменена.", "red"));
                return;
            }
        }

        var logLines = new List<string>();
        int deleted = 0;
        foreach (var d in emptyDirs)
        {
            if (verbose) Console.WriteLine(Colorize($"Удаление: {d}", "green"));
            try
            {
                Directory.Delete(d, true);
                deleted++;
                logLines.Add(d);
            }
            catch (Exception e)
            {
                Console.WriteLine(Colorize($"Ошибка удаления {d}: {e.Message}", "red"));
            }
        }

        if (logFile != null)
        {
            try
            {
                File.WriteAllLines(logFile, logLines);
                Console.WriteLine(Colorize($"Лог сохранён в {logFile}", "green"));
            }
            catch (Exception e)
            {
                Console.WriteLine(Colorize($"Ошибка записи лога: {e.Message}", "red"));
            }
        }

        Console.WriteLine(Colorize($"Удалено {deleted} папок.", "green"));
    }
}
